src/spec-lang/tc/typecheck.ts

```diff
@@ -1711,40 +1563,30 @@ export function tcFunctionCall(expr: SFunctionCall, ctx: STypingCtx, typeEnv: Ty
         return specializeType(calleeT.type, DataLocation.Memory);
     }
 
-    // Type-cast to a user-defined type
-    if (calleeT instanceof SUserDefinedTypeNameType) {
+    // Type-cast to a user-defined type or a struct constructor
+    if (calleeT instanceof UserDefinedType) {
         const argTs = expr.args.map((arg) => tc(arg, ctx, typeEnv));
-        let loc: DataLocation;
 
         if (calleeT.definition instanceof StructDefinition) {
-            // Struct constructor case
-            loc = DataLocation.Memory;
-        } else {
-            // Type-casting case
+            // Struct constructor case - always lives in memory
+            return new PointerType(mkUserDefinedType(calleeT.definition), DataLocation.Memory);
+        } else if (calleeT instanceof ContractDefinition || calleeT instanceof EnumDefinition) {
```

Comment:
This is buggy - the if condition should be `calleeT.definition instanceof ContractDefinition || calleeT.definition instanceof EnumDefinition`. Again its weird that tests didn't catch this. Perhaps we need more casting tests?

Reply:
Fixed in ef446bd.